model/project_event.go

```diff
@@ -0,0 +1,132 @@
+package model
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/evergreen-ci/evergreen/db"
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	ResourceTypeProject      = "PROJECT"
+	EventTypeProjectModified = "PROJECT_MODIFIED"
+	EventTypeProjectAdded    = "PROJECT_ADDED"
+)
+
+type ProjectSettings struct {
+	ProjectRef         ProjectRef           `bson:"proj_ref" json:"proj_ref"`
+	GitHubHooksEnabled bool                 `bson:"github_hooks_enabled" json:"github_hooks_enabled"`
+	Vars               ProjectVars          `bson:"vars" json:"vars"`
+	Aliases            []ProjectAlias       `bson:"aliases" json:"aliases"`
+	Subscriptions      []event.Subscription `bson:"subscriptions" json:"subscriptions"`
+}
+
+type ProjectChange struct {
+	User   string          `bson:"user" json:"user"`
+	Before ProjectSettings `bson:"before" json:"before"`
+	After  ProjectSettings `bson:"after" json:"after"`
+}
+
+// Embed the EventLogEntry so we can reimplement the SetBSON method for Projects
+type ProjectChangeEvent struct {
+	event.EventLogEntry
+}
+
+func (e *ProjectChangeEvent) SetBSON(raw bson.Raw) error {
+	temp := event.UnmarshalEventLogEntry{}
+	if err := raw.Unmarshal(&temp); err != nil {
+		return errors.Wrap(err, "can't unmarshal event container type")
+	}
+
+	e.Data = &ProjectChange{}
+
+	if err := temp.Data.Unmarshal(e.Data); err != nil {
+		return errors.Wrap(err, "failed to unmarshal data")
+	}
+
+	// IDs for events were ObjectIDs previously, so we need to do this
+	switch v := temp.ID.(type) {
+	case string:
+		e.ID = v
+	case bson.ObjectId:
+		e.ID = v.Hex()
+	default:
+		return errors.Errorf("unrecognized ID format for event %v", v)
+	}
+	e.Timestamp = temp.Timestamp
+	e.ResourceId = temp.ResourceId
+	e.EventType = temp.EventType
+	e.ProcessedAt = temp.ProcessedAt
+	e.ResourceType = temp.ResourceType
+
+	return nil
+}
+
+// Project Events queries
+func ProjectEventsForId(id string) db.Q {
```

Comment:
I know we do this in other places, but can we avoid having public functions that return db.Q objects, and instead just have public functions that return data?

Reply:
This seems to be the paradigm for events (events/event_finder.go). I could change it for projects, but it will be an outlier.

Just want to make sure.

Reply:
It is the paradigm, but db.Q is a fraught type.

Reply:
Done.